[style manager] improve UI

nirvn opened the pull request:
This PR does the following:
- create a dedicated set of buttons for addition of tags and smartgroups to make those actions visible, as well as get rid of the need to select a tag or a smartgroup to create those (quite confusing for newcomers, I've seen students I trained failing to understand this)
- move UI elements around, regroup add/remove/edit symbol buttons to harmonize with other parts of QGIS; the regrouping feels more natural now (or at least that's my hope)

The PR also switches the point,line,fill/polygon icon to rely of the vectorized icon QGIS uses elsewhere, as well as vectorizing the colorramp icon. That'll improve the situation with HiDPI screens ( @anitagraser :) )

Here's the before vs. proposed shot:
![untitled](https://cloud.githubusercontent.com/assets/1728657/20381251/36f0d21a-acd9-11e6-8955-8a7938997d9b.png)

NathanW2 commented:
A little tag and group icon for the tree items would be cool ;)

On Thu, 17 Nov 2016 6:04 pm Mathieu Pellerin <notifications@github.com>
wrote:

> This PR does the following:
>
>    - create a dedicated set of buttons for addition of tags and
>    smartgroups to make those actions visible, as well as get rid of the need
>    to select a tag or a smartgroup to create those (quite confusing for
>    newcomers, I've seen students I trained failing to understand this)
>    - move UI elements around, regroup add/remove/edit symbol buttons to
>    harmonize with other parts of QGIS
>
> The PR also switches the point,line,fill/polygon icon to rely of the
> vectorized icon QGIS uses elsewhere, as well as vectorizing the colorramp
> icon. That'll improve the situation with HiDPI screens ( @anitagraser
> <https://github.com/anitagraser> :) )
>
> Here's the before vs. proposed shot:
> [image: untitled]
> <https://cloud.githubusercontent.com/assets/1728657/20380856/1ab169f4-acd7-11e6-8c0d-191c7c8c2a9e.png>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/qgis/QGIS/pull/3778
> Commit Summary
>
>    - [style manager] improve UI
>
> File Changes
>
>    - *M* images/images.qrc
>    <https://github.com/qgis/QGIS/pull/3778/files#diff-0> (2)
>    - *M* python/gui/symbology-ng/qgsstylemanagerdialog.sip
>    <https://github.com/qgis/QGIS/pull/3778/files#diff-1> (6)
>    - *M* src/gui/symbology-ng/qgsstylemanagerdialog.cpp
>    <https://github.com/qgis/QGIS/pull/3778/files#diff-2> (138)
>    - *M* src/gui/symbology-ng/qgsstylemanagerdialog.h
>    <https://github.com/qgis/QGIS/pull/3778/files#diff-3> (6)
>    - *M* src/ui/qgsprojectpropertiesbase.ui
>    <https://github.com/qgis/QGIS/pull/3778/files#diff-4> (8)
>    - *M* src/ui/qgsstylemanagerdialogbase.ui
>    <https://github.com/qgis/QGIS/pull/3778/files#diff-5> (164)
>
> Patch Links:
>
>    - https://github.com/qgis/QGIS/pull/3778.patch
>    - https://github.com/qgis/QGIS/pull/3778.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/3778>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3AMbHlEHn0PrgK6rPcia8OBL0f8eks5q_AqJgaJpZM4K0-fd>
> .
>

DelazJ commented:
@nirvn, what about:
- adding also a label to the "remove tag/group" button. I find it a bit inconsistent to have labels for adding those items and none for their removal. This button will be clearly identified and no risk of confusion with the other minus button
- Unless you consider the import/export button as group-related tool (I do not recall all your recent improvements) and not a symbol related one, I'd suggest to move the symbols related buttons (add, remove, Edit, import/export) under the symbols preview frame (I think this will better help to figure out to which frame belong each set of tools.
- I'd add to these tools the "group symbols" button (which I can't make work on my 2.18 - I think it's broken given that it doesn't behave as described in http://docs.qgis.org/2.14/en/docs/user_manual/working_with_vector/style_library.html#groups-and-smart-groups). I suppose it is a kind of symbol edit (edit the tag property of several symbols at once) than a tag/group manager tool.

Thanks for your perpetual work on cleaning the UI and improving UX

nirvn commented:
@DelazJ ,
* regarding the import / export button, because the import and export functions targets all symbol categories at the same time (i.e. point, line, fill, and color ramp), I think it better sits where it is now. Plus, the export does make use of tag and smartgroup groupings. 
* as for moving the symbol related buttons, I'm not sure I get what your trying to express; are they not below the symbols preview frame already?
* again, you've lost me there, sorry. "these tools" refers to?

nirvn commented:
@DelazJ , oh, as for "remove tag / smart group" button, I tried to avoid adding a label as it would end up taking way too much horizontal space. I agree that it's not ideal as it stands next to another [ - ] button, but hopefully now that symbols preview list buttons vs. general actions buttons are separated visually, it's not too much of a problem.

anitagraser commented:
Thank you @nirvn! I think these changes make the dialog considerably clearer.

The only thing I'm uncertain about is the placement of the search field. I feel that it's usually on the top of the list rather than in the bottom right corner, e.g. in the Processing toolbox.

nirvn commented:
@DelazJ here's a screenshot of the import and export dialogs, which IMHO are pretty clearly going way beyond management of a single symbol type:
![untitled1](https://cloud.githubusercontent.com/assets/1728657/20384000/aa1404dc-ace4-11e6-9e9b-02e05d3394c3.png)

![untitled2](https://cloud.githubusercontent.com/assets/1728657/20384005/ae15ad06-ace4-11e6-93d0-1f3a5db614a8.png)

There's explicit links to the tags and smart groups, goes across symbols (and color ramps), etc.

DelazJ commented:
> as for moving the symbol related buttons, I'm not sure I get what your trying to express; are they not below the symbols preview frame already?

Yes they are! I meant that if the import/export tool was considered as symbol related, it'd be nice to have it at the place of the plus button and offset the tools to the right. But as it works with group, ignore the comment, it's ok.
> again, you've lost me there, sorry. "these tools" refers to?

The symbol-related tools I meant, sorry. My idea is to move this tool near the plus, minus, edit buttons of symbols (though I don't know how it currently works)

What about `Remove tag/group` label for the button? we no longer need to mention smart as simple groups do not exist.

nirvn commented on src/gui/symbology-ng/qgsstylemanagerdialog.cpp:
```diff
@@ -167,8 +174,12 @@ QgsStyleManagerDialog::QgsStyleManagerDialog( QgsStyle* style, QWidget* parent )
   mGroupTreeContextMenu->addAction( actnEditSmartGroup );
-  connect( actnAddGroup, SIGNAL( triggered( bool ) ), this, SLOT( addGroup() ) );
-  mGroupTreeContextMenu->addAction( actnAddGroup );
+  connect( actnAddTag, &QAction::triggered, [=]( bool t ) { this->addTag(); }
```
@nyalldawson , lambda in connect! ;)

nirvn commented:
@anitagraser , I could move it up again, except that it felt like a great way to save vertical spacing and leave more space for the star of this dialog: the symbols list. Should we agree to test this relocation and see how it goes? If you feel it irritates you too much, we can always push it back up later (although I'd shed a tear ;) )

nirvn commented:
@DelazJ , "group" just doesn't really what the smart group is all about, don't you think? If I would see "group", I'd just think of the old (RIP) group and ignore it. Smart groups let user associate that function with itune's smart playlist, etc.

nirvn commented:
@DelazJ , actually, there is an alternative: we could simply get rid of that button, and have tags and smart groups removal done only via right-click:
![untitled2](https://cloud.githubusercontent.com/assets/1728657/20384466/8b1a55ca-ace6-11e6-86d1-3745f2e7d84e.png)

DelazJ commented:
> @DelazJ , actually, there is an alternative: we could simply get rid of that button, and have tags and smart groups removal done only via right-click

Wouldn't that be one click more?
Another helper could be a tooltip on what it really does (mentionning smart in it). No strong opinion

nirvn commented:
Ok, discussing some more with @DelazJ , we settled on this:
![untitled](https://cloud.githubusercontent.com/assets/1728657/20385822/92a8ca6a-aceb-11e6-9614-60d3a1d0a337.png)

Commit updated.

nyalldawson commented:
Hate to come to the discussion late, but I'm not a fan of the text-heavy buttons in the bottom left. IMO this detracts from the clean looks of the dialog and removes the focus from where it should be - the symbols themselves. My vote would go to just having the remove options in a right click menu. I actually think this is better UX as there's not much association between click a tag/group at the top of that list and the remove button way down the bottom. 

Some other things I think could greatly improve this dialog would be:
- If i right click a symbol and select "Add to tag" , there should be a "Create new tag..." option listed. 
- It'd be very nice to see counts of symbols next to the items in the list on the left. Eg "Favourites (5)", "My tag (15)", etc.

nyalldawson commented:
By the way - on master I can't add symbols to the favourites group via right click - "add to favourites". Nothing happens.

nyalldawson commented:
Another bit of polish:
- the "type here to filter symbols" text should change to "type here to filter ramps" when the color ramp tab is active

nirvn committed:
[style manager] improve UI
- create a dedicated set of buttons for addition of tags and smartgroups
to make those actions more visible as well as getting rid of th need to
select a tag/smartgroup to create those (turns out to be quite confusing
for newcomers)
- move UI elements around, regroup {add,remove,edit} symbol buttons to
harmonize with other parts of QGIS; the elements' placement feels much
more natural now

nirvn commented:
@nyalldawson , commit updated with the following points addressed:
* Ok, I've removed the "Remove tag/group" button in favor of the right-click solution; while re-thinking about this overnight, I became less certain of the above-proposed button solution, glad you've chipped in on this one
* A "Create new tag..." option was added in the right-click menu, great idea.
* The search box placeholder now updates to reflect symbol vs color ramp, good catch

I'm not super enthusiastic about the number (X) attached to the tag name; it'd have to be updated every time you switch the select tab (i.e. point, line, fill, color ramp), etc. 

And thanks for jumping into the discussion, feedback and ideas always great to have.

nirvn committed:
[style manager] sort displayed symbols, insure tags are added only once

nyalldawson commented:
@nirvn can you post some screenshots of the current version?

nirvn commented:
@nyalldawson, adding test case as we speak.

Here's a screenshot of latest version:
![untitled](https://cloud.githubusercontent.com/assets/1728657/20419474/3c898bb2-ad89-11e6-954f-ed00252df104.png)

nirvn committed:
tests for two additional QgsStyle functions: tag() & symbolHasTag()

nirvn commented:
@nyalldawson , tests added for tag() as well as symbolHasTag() -- the latter was added after I stumbled on an issue whereas a symbol can get tagged with the same tag many times, making it appear twice (or more) in a tag list.

nirvn commented:
@nyalldawson , screenshot of the "create new tag..." menu shortcut:
![untitled](https://cloud.githubusercontent.com/assets/1728657/20420538/0e28f192-ad91-11e6-9b13-723e0f151bfb.png)

nyalldawson commented:
Sweet stuff.. this polish is very nice!
